velox/exec/StreamingWindow.cpp

```diff
@@ -0,0 +1,199 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "velox/exec/StreamingWindow.h"
+#include "velox/exec/OperatorUtils.h"
+#include "velox/exec/Task.h"
+
+namespace facebook::velox::exec {
+
+void StreamingWindow::updateValues() {
+  partitionStartRows_.push_back(sortedRows_.size());
+  sortedRows_.insert(
+      sortedRows_.end(), wholeGroupedRows_.begin(), wholeGroupedRows_.end());
+  wholeGroupedRows_.clear();
+}
+
+void StreamingWindow::addInput(RowVectorPtr input) {
+  for (auto col = 0; col < input->childrenSize(); ++col) {
+    decodedInputVectors_[col].decode(*input->childAt(col));
+  }
+
+  auto partitionCompare = [&](const char* lhs, const char* rhs) -> bool {
+    return compareRowsWithKeys(lhs, rhs, partitionKeyInfo_);
+  };
+
+  // Add all the rows into the RowContainer.
+  for (auto row = 0; row < input->size(); ++row) {
+    char* newRow = data_->newRow();
+
+    for (auto col = 0; col < input->childrenSize(); ++col) {
+      data_->store(decodedInputVectors_[col], row, newRow, col);
+    }
+
+    if (preRow_ != nullptr && partitionCompare(preRow_, newRow)) {
+      // the wholeGroupedRows_ already store whole group, we need to put the
+      // wholeGroupedRows_ into the sortedRows and then update the
+      // partitionStartRows_ value.
+      updateValues();
+    }
+
+    wholeGroupedRows_.push_back(newRow);
+    preRow_ = newRow;
+  }
+
+  partitionStartRows_.push_back(sortedRows_.size());
+
+  outputBatchSize_ = outputBatchRows(data_->estimateRowSize());
+}
+
+inline bool StreamingWindow::compareRowsWithKeys(
```

Comment:
Can we make this a function in the parent class that can be derived from ?

Reply:
Yes.